safekeeper/src/wal_storage.rs

```diff
@@ -509,10 +518,23 @@ impl WalReader {
             wal_segment: None,
             enable_remote_read,
             local_start_lsn: state.local_start_lsn,
+            timeline_start_lsn: state.timeline_start_lsn,
         })
     }
 
     pub async fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        // If this timeline is new, we may not have a full segment yet, so
+        // we pad the first bytes of the timeline's first WAL segment with 0s
```

Comment:
Could you remind me how it works? So, PG walreceiver always starts streaming at the beginning of the segment, despite startup process starting redo where it should, i.e. basebackup lsn in our case? It's a bit strange to send garbage zeros, but seems like it is optimal here.

Reply:
Let's instead perform this initialization in `write_wal`, when `self.write_lsn` is invalid, i.e. on first write. That way we don't need temp in memory segment, and segment in safekeepers will be pg_waldump'able.

Reply:
> In addition, I think that the temporary extra RAM usage for replicas at the Safekeeper doesn't weigh up to the permanent extra storage required for timeline start segments.

I agree that RAM usage is not an issue, but neither I think storage usage is (BTW we explicitly set all 16MB to zeros). But as I said pg_waldumpability is really nice, and initializing properly (kinda) in the first place is more natural, you don't need this byte fiddling joining synthetic part and real one. Existing data is an issue indeed; I'd prefer to write a patch like Arthur did [here](https://github.com/neondatabase/neon/pull/3916) to fix them. I don't strongly insist though, I can do that later.